tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,219 @@
+        return asyncio.run(self.store.exists(key))
+
+    def list_dir(self, prefix):
+        raise NotImplementedError
+
+    def list_prefix(self, prefix: str):
+        raise NotImplementedError
+
+    def set_partial_values(self, key_start_values):
+        raise NotImplementedError
+
+    def supports_listing(self):
+        raise NotImplementedError
+
+    def supports_partial_writes(self):
+        raise NotImplementedError
+
+    def supports_writes(self):
+        raise NotImplementedError
+
+
+class ZarrStoreStateMachine(RuleBasedStateMachine):
+    def __init__(self):
+        super().__init__()
+        self.model: dict[str, bytes] = {}
+        self.store = SyncStoreWrapper(MemoryStore(mode="w"))
+        self.prototype = default_buffer_prototype()
+
+    @rule(key=paths, data=st.binary(min_size=0, max_size=100))
+    def set(self, key: str, data: bytes) -> None:
+        note(f"(set) Setting {key!r} with {data}")
+        assert not self.store.mode.readonly
+        data_buf = Buffer.from_bytes(data)
+        self.store.set(key, data_buf)
+        self.model[key] = data_buf
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(key=paths, data=st.data())
+    def get(self, key, data) -> None:
+        key = data.draw(
+            st.sampled_from(sorted(self.model.keys()))
+        )  # hypothesis wants to sample from sorted list
+        model_value = self.model[key]
+        note(f"(get) model value {model_value.to_bytes()}")
+        store_value = self.store.get(key, self.prototype)
+        note(f"(get) store value: {store_value.to_bytes()}")
+        # to bytes here necessary because data_buf set to model in set()
+        assert self.model[key].to_bytes() == (store_value.to_bytes())
+
+    @rule(key=paths, data=st.data())
+    def get_invalid_keys(self, key, data) -> None:
+        model_keys = list(self.model.keys())
+        # model_keys = ['/']
+        # key = '/'
+        note("(get_invalid)")
+
+        # note(f"(get invalid) key: {key}")
+        # note(f"(get invalid) val: {self.store.get(key, self.prototype)}")
+
+        assume(key not in model_keys)
+        assert self.store.get(key, self.prototype) is None
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(data=st.data())
+    def get_partial_values(self, data) -> None:
+        key_range = data.draw(key_ranges(keys=st.sampled_from(sorted(self.model.keys()))))
+        note(f"(get partial) {key_range=}")
+        obs_maybe = self.store.get_partial_values(key_range, self.prototype)
+        observed = []
+
+        for obs in obs_maybe:
+            assert obs is not None
+            observed.append(obs.to_bytes())
+
+        model_vals_ls = []
+
+        for key, byte_range in key_range:
+            model_vals = self.model[key]
+            start = byte_range[0] or 0
+            step = byte_range[1]
+            stop = start + step if step is not None else None
+            model_vals_ls.append(model_vals[start:stop])
```

Comment:
```suggestion
            start = byte_range[0] or 0
            step = byte_range[1]
            stop = start + step if step is not None else None
            model_vals_ls.append(self.model[key][start:stop])
```

Reply:
thanks! adding this and the get invalid change locally